synapse/dendrite/jobs.py

```diff
@@ -0,0 +1,214 @@
+        if tufo:
+            # for now, just log and requeue the job
+            logger.debug('Failing job %s', tufo)
+            self.core.setTufoProps(tufo, status='failed', failed_at=self._currTimestamp())
+
+    def qsize(self, queue, status='queued'):
+        '''
+        Determine the size of the queue specified.
+
+        By default, the number of 'queued' jobs is returned. A status is optionally specified to determine the number
+        of jobs in other states, e.g. 'working' or 'completed'.
+
+        Args:
+            queue (str):
+                The name of the queue to determine the size of.
+
+            status (str):
+                The job status to predicate on while determining the size.
+
+        Returns:
+            int: Size of the queue.
+        '''
+        return len(self._jobsByQueue(queue, status))
+
+    def isEmpty(self, queue, status='queued'):
+        '''
+        Determine whether the queue specified is empty.
+
+        A convenience method for determining whether a queue's size is 0.
+
+        Args:
+            queue (str):
+                The name of the queue to determine the size of.
+
+            status (str):
+                The job status to predicate on while determining the size.
+
+        Returns:
+            bool: True if the queue is empty, False otherwise.
+        '''
+        return self.qsize(queue, status) == 0
+
+    def stats(self, status='any'):
+        '''
+        Returns a histogram showing job count by queue name.
+
+        The default status argument is 'any', which is a pseudo status for collecting all jobs regardless of status.
+
+        Args:
+            status (str):
+                The status to filter the jobs by, e.g. 'queued', 'working', 'completed'.
+
+        Returns:
+            collections.defaultdict: A dictionary containing a histogram of count by queue name.
+        '''
+        if status == 'any':
+            return self.core.getStatByProp('histo', 'dendrite:job:queue')
+
+        histo = collections.defaultdict(int)
+        for row in self.core.getJoinByProp('dendrite:job:status', status):
+            if row[1] == 'dendrite:job:queue':
+                histo[row[2]] += 1
+        return histo
+
+    def _buildJobTufo(self, queue, job):
+        return {
+            'queue':    queue,
+            'data':     json.dumps(job),
+            'runkey':   self._runkey(),
+            'status':   'queued',
+            'queued_at': self._currTimestamp()
+        }
+
+    def _runkey(self):
+        currms = self._currTimestamp()
+        if currms != self.currMs:
+            self.currMs = currms
+            self.jobNum = 1
+        else:
+            self.jobNum += 1
+        return (currms << 8) | self.jobNum
+
+    def _currTimestamp(self):
+        return int(time.time() * 1000)
+
+    def _jobsByQueue(self, queue, status='queued'):
+        jobs = []
+        for tufo in self.core.getTufosByProp('dendrite:job:queue', queue):
```

Comment:
If i'm not mistaken, this logic will lift all the jobs that have ever existed for a given queue in order to loop filter them down to the ones with the given state.  After running long term, this will literally lift millions of jobs from ages ago to ultimately return just a few...